---
title: "Comparison social forms *Solenopsis invcita* using the Morandin et al., 2016 data"
author: "Carlos Martinez Ruiz"
date: "31 October 2018"
output: pdf_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.path = 'results/')
```
```{r libraries, echo = FALSE, message = FALSE}
#Load all libraries
library(tximport)
library(readr)
library(DESeq2)
library(GenomicRanges)
library(GenomicFeatures)
library(ggplot2)
for (package in (.packages()) ) { 
 print(paste("Package", package, "version", package.version(package)))
}

```

## Introduction
The (Morandin et al., 2016)<https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0902-7> data consits in 3 samples from workers and queens from the two social forms of *Solenopsis invicta*. The aim of this analysis is to obtain differentially expressed genes between social forms in queens and workers independently and check whether or not the social chromosome is enriched in DEGs. The counts used for this analysis were generated by Kalisto (v0.43.0), using the available *S.invicta* coding sequences (gnG assembly, version GCF_000188075.1).
Two samples were removed from the analysis due to poor alignment, a polygyne queen and a monogyne worker. 

## Import data from Kallisto into R using 'tximport'

The data from Kallisto first needs to be loaded into R, removing any problematic samples. In this case two samples were removed, a polygyne queen (sample DRR042171) and a monogyne worker (DRR042169), due to poor alignment to the reference. Tha samples then have a more workable name assigned and stored into colData, a matrix that will be then used in DESeq. This matrix assigns factors to each sample, in this case, caste and social form. Both the names and the assigned factors were hard coded, so it is worth checking the table once generated. 

```{r load_data_r, message = FALSE}
# Load Kallisto abundance files

paths <- list()
paths$data <- "input"
paths$kallisto_output <- file.path(paths$data)

paths$kallisto_files_relative <- grep(x       = list.files(paths$kallisto_output,
                                                           recursive = TRUE),
                                      pattern = ".tsv",
                                      value   = TRUE)

paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)



#Extract the sample names:
samples <- gsub(x           = paths$kallisto_files_relative,
                pattern     = "/.*",
                replacement = "")

#Label the files:
names(paths$kallisto_files) <- samples
#Generate the colData table for DESeq2 analysis:
soc_form <- gsub(x = samples, pattern = "_.+", replacement = "")
caste    <- gsub(x = samples, pattern = "(.+_)([QW])([0-9])", replacement = "\\2")

colData           <- matrix(c(samples, caste, soc_form), ncol = 3)
rownames(colData) <- samples
colnames(colData) <- c("sample", "caste", "social_form")

#Remove problematic samples Poly_Q1 and Mono_W2
colData <- colData[-grep("Poly_Q1|Mono_W2", colData[, 1]), ]

paths$kallisto_files <- paths$kallisto_files[-grep("Poly_Q1|Mono_W2",
                                                   paths$kallisto_files)]
#Check the resulting matrix
colData
```

Once the data has been loaded into R, it needs to be ran through tximport to make it compatible with DESeq2. The analysis in DESeq2 needs to be performed independently in workers and queens, so the data will be divided in two.The analysis will be performed in genes rather than transcripts (the original input). Tximport will assign the counts per transcript to their corresponding genes. For this to work, a table with each transcript and its corresponding gene needs to be loaded into R. This table was generated based on the gene annotation from RefSeq (gnG assembly, version GCF_000188075.1). 

```{r _load_data_tximport, message = FALSE, warning = FALSE}

#Separate data into workers and queens:
info <- list()
info$w <- colData[colData[, "caste"] == "W", ]
info$q <- colData[colData[, "caste"] == "Q", ]

kallisto_files_w <- paths$kallisto_files[rownames(info$w)]
kallisto_files_q <- paths$kallisto_files[rownames(info$q)]

#Load a table with the transcript ID and the gene to which they belong.

si_ann <- makeTxDbFromGFF(file = "input/GCF_000188075.1_Si_gnG_genomic.gff",
                          format="gff3")

#Generate a table with the gene names and their associated transcripts
gene_ids <- keys(si_ann, "GENEID")
tx_gene <- select(x = si_ann,
                  keys = gene_ids,
                  columns = c("TXNAME", "GENEID"),
                  keytype = "GENEID")

tx_gene <- as.data.frame(cbind(tx_gene$TXNAME, tx_gene$GENEID))
colnames(tx_gene) <- c("transcript", "gene")

#Import the abundance files into a DESeq2 object

txi_reads_w <- tximport(kallisto_files_w, type = "kallisto", reader = read_tsv,
                        tx2gene = tx_gene)
txi_reads_q <- tximport(kallisto_files_q, type = "kallisto", reader = read_tsv,
                        tx2gene = tx_gene)
```

## Perform the DESeq2 analysis

Once the data is ready, the analysis with DESeq2 can be performed by simply running `DESeq()`. **NOTE:** Because this version of DESeq2 is prior to the 1.16 update, the LFCs are automatically shrinked. In newer versions, `betaPrior = TRUE` needs to be added to the `DESeq()` function to get the same results. 

```{r deseq2_analysis, message = FALSE}
#FOR QUEENS:
ddsTxi_q <- DESeqDataSetFromTximport(txi_reads_q, colData = info$q,
                                     design = ~ social_form)

#Performs DESeq2 test for detecting DE loci
de_txi_q <- DESeq(ddsTxi_q)

#summary of the results:
res_de_txi_q <- results(de_txi_q)

#FOR WORKERS:
ddsTxi_w <- DESeqDataSetFromTximport(txi_reads_w, colData = info$w,
                                     design = ~ social_form)

#Performs DESeq2 test for detecting DE loci
de_txi_w <- DESeq(ddsTxi_w)

#summary of the results:
res_de_txi_w <- results(de_txi_w)
#Save the results
res_q_name <- paste0("results/", Sys.Date(),
                     "_results_deseq2_queens_mor.RData")
save(res_de_txi_q, file = res_q_name)
res_w_name <- paste0("results/", Sys.Date(),
                     "_results_deseq2_workers_mor.RData")
save(res_de_txi_w, file = res_w_name)
```

How many loci were detected by DESeq as differentially expressed?

**In workers**
```{r}
sum(res_de_txi_w$padj < 0.05, na.rm = TRUE)
```

**In queens**
```{r}
sum(res_de_txi_q$padj < 0.05, na.rm = TRUE)
```
**In common**

```{r}
de_genes_queens <- rownames(res_de_txi_q[which(res_de_txi_q$padj < 0.05), ])
de_genes_workers <- rownames(res_de_txi_w[which(res_de_txi_w$padj < 0.05), ])

de_genes_both <- de_genes_queens[de_genes_queens %in% de_genes_workers]
length(de_genes_both)
```
** Direction of expression **
```{r}
#In queens
de_genes_queens_poly <- rownames(res_de_txi_q[which(res_de_txi_q$padj < 0.05 &
                                                    res_de_txi_q$log2FoldChange > 0), ])

de_genes_queens_mono <- rownames(res_de_txi_q[which(res_de_txi_q$padj < 0.05 &
                                                    res_de_txi_q$log2FoldChange < 0), ])
#In workers
de_genes_workers_poly <- rownames(res_de_txi_w[which(res_de_txi_w$padj < 0.05 &
                                                    res_de_txi_w$log2FoldChange > 0), ])

de_genes_workers_mono <- rownames(res_de_txi_q[which(res_de_txi_w$padj < 0.05 &
                                                    res_de_txi_w$log2FoldChange < 0), ])

#For genes in common 
#In queens
res_de_txi_q_common <- res_de_txi_q[de_genes_both, ]

de_genes_queens_poly_common <- rownames(res_de_txi_q_common[which(res_de_txi_q_common$padj < 0.05 &
                                                    res_de_txi_q_common$log2FoldChange > 0), ])

de_genes_queens_mono_common <- rownames(res_de_txi_q_common[which(res_de_txi_q_common$padj < 0.05 &
                                                    res_de_txi_q_common$log2FoldChange < 0), ])

#In workers
res_de_txi_w_common <- res_de_txi_w[de_genes_both, ]

de_genes_workers_poly_common <- rownames(res_de_txi_w_common[which(res_de_txi_w_common$padj < 0.05 &
                                                    res_de_txi_w_common$log2FoldChange > 0), ])

de_genes_workers_mono_common <- rownames(res_de_txi_w_common[which(res_de_txi_w_common$padj < 0.05 &
                                                    res_de_txi_w_common$log2FoldChange < 0), ])
```

Polygyne-biased genes in queens:
```{r}
length(de_genes_queens_poly)
```
Monogyne-biased genes in queens:
```{r}
length(de_genes_queens_mono)
```
Binomial test
```{r}
binom.test(length(de_genes_queens_mono), 
           (length(de_genes_queens_mono) + length(de_genes_queens_poly)))
```
Binomial test for genes in common only
```{r}
binom.test(length(de_genes_queens_mono_common), 
           (length(de_genes_queens_mono_common) + length(de_genes_queens_poly_common)))
```


Polygyne-biased genes in workers:
```{r}
length(de_genes_workers_poly)
```
Monogyne-biased genes in workers:
```{r}
length(de_genes_workers_mono)
```
Binomial test
```{r}
binom.test(length(de_genes_workers_mono), 
           (length(de_genes_workers_mono) + length(de_genes_workers_poly)))
```

Binomial test for genes in common only
```{r}
binom.test(length(de_genes_workers_mono_common), 
           (length(de_genes_workers_mono_common) + length(de_genes_workers_poly_common)))
```

Plot the results in an MA plot, where the y axis is the LFCs between Polygyne and Monogyne individuals (i.e., positive values imply higher expression in polygyne colonies and vice-versa) and the x axis is the mean read count per gene. Red dots indicate genes which have been detected as significant (alpha = 0.05) according to DESeq. The differences between social forms are plotted independently for queens and workers.
```{r maplot}
par(mfrow = c(1, 2), mar = c(5, 4.6, 4, 2) + 0.1)
plotMA(res_de_txi_w, main = "Workers", ylim = c(-10, 10), xlab = "Mean normalized counts",
       ylab = "Logarithm Fold Change", font.lab = 1, cex.axis = 1, cex.lab = 1,
       cex.main = 2, cex = 0.5)
plotMA(res_de_txi_q, main = "Queens", ylim = c(-10, 10), xlab = "Mean normalized counts",
       ylab = "Logarithm Fold Change", font.lab = 1, cex.axis = 1, cex.lab = 1,
       cex.main = 2, cex = 0.5)

```

## Genome location analysis

Once the differentially expressed genes between social forms have been detected, the next question to answer is whether there is an enrichment of socially biased loci in the supergene region. 
For this, a table with the position of the supergene in the *Solenopsis invicta* genome (gnG assembly) needs to be loaded. The overlap between the position of the supergene and the location of the differentially expressed genes in then calculated using the packages GenomeRanges and IRanges. 
First step, get the positions of the supergene and of all the genes in the genome into R

```{r load_postions, warning = FALSE, message = FALSE}
#Load the file containing the last known position of the inversion in
#the S.invicta genome (for gnG genome ref)
regions <- read.table("input/gng_regions.tab", header = TRUE)

#Load annotation file 
annot <- makeTxDbFromGFF(file = "input/GCF_000188075.1_Si_gnG_genomic.gff")
#Returns coding regions by transcript (option "gene", otherwise it would extract features 
#(transcripts in this case) by transcript using the option "tx").
ann_ranges <- transcriptsBy(annot, "gene")
ann_ranges <- unlist(ann_ranges)
#Create a GRanges object from the 'regions' table
regions_ranges <- GRanges(Rle(regions$scaffold), IRanges(start = regions$start,
                                                         end = regions$end),
                          lg = regions$lg, region = regions$region, linkage_group = regions$lg)

#Create a GRanges object intersecting each gene to its corresponding region
gene_region <- mergeByOverlaps(regions_ranges, ann_ranges, ignore.strand = TRUE)

#Parse the 'gene_region' GRanges object to make it useful
loc_genes_region           <- data.frame(gene_region@listData$region,
                                         gene_region@listData$linkage_group,
                                         gene_region@listData$ann_ranges@ranges@NAMES)
colnames(loc_genes_region) <- c("region", "linkage_group", "loc_id")
loc_genes_region           <- unique(loc_genes_region)
```

Once the position of all genes within the genome (within or outside of the supergene) has been established, we can finally calculate the number of socially biased loci (defined as those detected as differentially expressed between social forms by DESeq2) within and outside the supergene. It would be also interesting to look at the number of genes significantly more highly expressed in polygyne (and vice-versa). All analyses are performed in queens and workers independently. 

```{r extract_numbers_of_socially_biased_genes}
#Make sure that only genes present in the annotation file are present
# in the DEseq2.
res_de_w <- res_de_txi_w[rownames(res_de_txi_w) %in% loc_genes_region$loc_id, ]
res_de_q <- res_de_txi_q[rownames(res_de_txi_q) %in% loc_genes_region$loc_id, ]

#Get gene names of DE loci in workers and queens
de_loci_w <- rownames(res_de_w[which(res_de_w$padj < 0.05), ])
de_loci_q <- rownames(res_de_q[which(res_de_q$padj < 0.05), ])

#Get gene names of loci over-expressed in polygyne workers and queens
overp_loci_w <- rownames(res_de_w[which(res_de_w$log2FoldChange > 0 &
                                                 res_de_w$padj < 0.05), ])
overm_loci_w <- rownames(res_de_w[which(res_de_w$log2FoldChange < 0 &
                                                 res_de_w$padj < 0.05), ])

overp_loci_q <- rownames(res_de_q[which(res_de_q$log2FoldChange > 0 &
                                                 res_de_q$padj < 0.05), ])
overm_loci_q <- rownames(res_de_q[which(res_de_q$log2FoldChange < 0 &
                                                 res_de_q$padj < 0.05), ])

#Add columns for de genes in workers and queens to the regions df
loc_genes_region$de_w <- "non-significant"
loc_genes_region$de_q <- "non-significant"

#Label de genes in workers and queens
loc_genes_region[match(de_loci_q, loc_genes_region$loc_id), ]$de_q <- "significant"
loc_genes_region[match(de_loci_w, loc_genes_region$loc_id), ]$de_w <- "significant"

#Add columns the direction of over-expression in workers and queens
loc_genes_region$expr_dir_q <- "Non-diff"
loc_genes_region$expr_dir_w <- "Non-diff"

#Label over expressed genes in polygyne workers and queens
loc_genes_region[match(overp_loci_q, loc_genes_region$loc_id), ]$expr_dir_q <- "Poly"
loc_genes_region[match(overp_loci_w, loc_genes_region$loc_id), ]$expr_dir_w <- "Poly"

loc_genes_region[match(overm_loci_q, loc_genes_region$loc_id), ]$expr_dir_q <- "Mono"
loc_genes_region[match(overm_loci_w, loc_genes_region$loc_id), ]$expr_dir_w <- "Mono"

#Workers
de_rec_w <- length(which(loc_genes_region$region == "recombining" &
                           loc_genes_region$de_w == "significant"))

non_de_rec_w <- length(which(loc_genes_region$region == "recombining" &
                               loc_genes_region$de_w == "non-significant"))

de_non_rec_w <- length(which(loc_genes_region$region == "supergene" &
                               loc_genes_region$de_w == "significant"))

non_de_non_rec_w <- length(which(loc_genes_region$region == "supergene" &
                                   loc_genes_region$de_w == "non-significant"))

#Queens
de_rec_q <- length(which(loc_genes_region$region == "recombining" &
                           loc_genes_region$de_q == "significant"))

non_de_rec_q <- length(which(loc_genes_region$region == "recombining" &
                               loc_genes_region$de_q == "non-significant"))

de_non_rec_q <- length(which(loc_genes_region$region == "supergene" &
                               loc_genes_region$de_q == "significant"))

non_de_non_rec_q <- length(which(loc_genes_region$region == "supergene" &
                                   loc_genes_region$de_q == "non-significant"))
```

Generate a barplot showing both the distribution of socially biased loci within or outside the supergene region and the porportion of highly to lowly expressed loci in the different social forms

```{r socialy_biased_loci_in_the_genome}
#####Barplot to show proportion of socially biased genes within and
# outside the supergene region ######


de_rec_m_q <- length(which(loc_genes_region$region == "recombining" &
                             loc_genes_region$de_q == "significant" &
                             loc_genes_region$expr_dir_q == "Mono"))

de_rec_p_q <- length(which(loc_genes_region$region == "recombining" &
                             loc_genes_region$de_q == "significant" &
                             loc_genes_region$expr_dir_q == "Poly"))

non_de_rec_m_q <- length(which(loc_genes_region$region == "recombining" &
                                 loc_genes_region$de_q == "non-significant" &
                                 loc_genes_region$expr_dir_q == "Mono"))

non_de_rec_p_q <- length(which(loc_genes_region$region == "recombining" &
                                 loc_genes_region$de_q == "non-significant" &
                                 loc_genes_region$expr_dir_q == "Poly"))

de_non_rec_m_q <- length(which(loc_genes_region$region == "supergene" &
                                 loc_genes_region$de_q == "significant" &
                                 loc_genes_region$expr_dir_q == "Mono"))

de_non_rec_p_q <- length(which(loc_genes_region$region == "supergene" &
                                 loc_genes_region$de_q == "significant" &
                                 loc_genes_region$expr_dir_q == "Poly"))

non_de_non_rec_m_q <- length(which(loc_genes_region$region == "supergene" &
                                     loc_genes_region$de_q == "non-significant" &
                                     loc_genes_region$expr_dir_q == "Mono"))

non_de_non_rec_p_q <- length(which(loc_genes_region$region == "supergene" &
                                     loc_genes_region$de_q == "non-significant" &
                                     loc_genes_region$expr_dir_q == "Poly"))

#Barplot for queens
de_percent_nrec_m_q <- de_non_rec_m_q / (non_de_non_rec_q + de_non_rec_q) * 100
de_percent_nrec_p_q <- de_non_rec_p_q / (non_de_non_rec_q + de_non_rec_q) * 100
de_percent_rec_m_q <- de_rec_m_q / (de_rec_q + non_de_rec_q) * 100
de_percent_rec_p_q <- de_rec_p_q / (de_rec_q + non_de_rec_q) * 100
expected_percent_q <- (de_non_rec_q + de_rec_q) / (non_de_non_rec_q + non_de_rec_q) * 100

to_plot_q           <- data.frame(c(de_percent_nrec_m_q, de_percent_nrec_p_q,
                                    de_percent_rec_m_q, de_percent_rec_p_q),
                                  rep(c("Supergene", "Rest of the genome"), each = 2),
                                  rep(c("Mono", "Poly"), n = 2))
colnames(to_plot_q) <- c("percentage", "region", "form")


p1 <- ggplot(data = to_plot_q, aes(x = region, y = percentage, fill = form)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = expected_percent_q, colour = "red") +
  theme_bw() + theme(panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
                     legend.title = element_blank()) +
 scale_x_discrete(name = "") + scale_y_continuous(name = "% differentially expressed genes",
                                                   limits = c(0, 8)) +
  scale_fill_manual(values = c("brown4", "cornflowerblue"), labels = c("Single-queen\nbias", "Multiple-queen\nbias")) +
  theme(axis.text.x = element_text(size = 16), axis.text.y = element_text(size = 16),
        axis.title.y = element_text(size = 20), legend.text = element_text(size = 16))


#Same plot, for workers
de_rec_m_w <- length(which(loc_genes_region$region == "recombining" &
                             loc_genes_region$de_w == "significant" &
                             loc_genes_region$expr_dir_w == "Mono"))

de_rec_p_w <- length(which(loc_genes_region$region == "recombining" &
                             loc_genes_region$de_w == "significant" &
                             loc_genes_region$expr_dir_w == "Poly"))

non_de_rec_m_w <- length(which(loc_genes_region$region == "recombining" &
                                 loc_genes_region$de_w == "non-significant" &
                                 loc_genes_region$expr_dir_w == "Mono"))

non_de_rec_p_w <- length(which(loc_genes_region$region == "recombining" &
                                 loc_genes_region$de_w == "non-significant" &
                                 loc_genes_region$expr_dir_w == "Poly"))

de_non_rec_m_w <- length(which(loc_genes_region$region == "supergene" &
                                 loc_genes_region$de_w == "significant" &
                                 loc_genes_region$expr_dir_w == "Mono"))

de_non_rec_p_w <- length(which(loc_genes_region$region == "supergene" &
                                 loc_genes_region$de_w == "significant" &
                                 loc_genes_region$expr_dir_w == "Poly"))

non_de_non_rec_m_w <- length(which(loc_genes_region$region == "supergene" &
                                     loc_genes_region$de_w == "non-significant" &
                                     loc_genes_region$expr_dir_w == "Mono"))

non_de_non_rec_p_w <- length(which(loc_genes_region$region == "supergene" &
                                     loc_genes_region$de_w == "non-significant" &
                                     loc_genes_region$expr_dir_w == "Poly"))

de_percent_nrec_m_w <- de_non_rec_m_w / (non_de_non_rec_w + de_non_rec_w) * 100
de_percent_nrec_p_w <- de_non_rec_p_w / (non_de_non_rec_w + de_non_rec_w) * 100
de_percent_rec_m_w <- de_rec_m_w / (de_rec_w + non_de_rec_w) * 100
de_percent_rec_p_w <- de_rec_p_w / (de_rec_w + non_de_rec_w) * 100
expected_percent_w <- (de_non_rec_w + de_rec_w) / (non_de_non_rec_w + non_de_rec_w) * 100

to_plot_w           <- data.frame(c(de_percent_nrec_m_w, de_percent_nrec_p_w,
                                    de_percent_rec_m_w, de_percent_rec_p_w),
                                  rep(c("Supergene", "Rest of the genome"), each = 2),
                                  rep(c("Mono", "Poly"), n = 2))
colnames(to_plot_w) <- c("percentage", "region", "form")


p2 <- ggplot(data = to_plot_w, aes(x = region, y = percentage, fill = form)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = expected_percent_w, colour = "red") +
  theme_bw() + theme(panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
                     legend.title = element_blank()) +
  scale_x_discrete(name = "") + scale_y_continuous(name = "% of differentially expressed genes\nbetween social forms",
                                                   limits = c(0, 8)) +
  scale_fill_manual(values = c("brown4", "cornflowerblue"), labels = c("Single-queen\nbias", "Multiple-queen\nbias")) +
  theme(axis.text.x = element_text(size = 16), axis.text.y = element_text(size = 16),
        axis.title.y = element_text(size = 20))
```

The red lines in the plots show the proportion of socially biased loci expected by randomness in workers or queens. The blue portion of each bar represents the proportion of those socially biased loci that are more highly expressed in polygyne individuals and vice-versa for the burgundy portion.
Plot for queens
```{r, echo = FALSE}
p1
```

Plot for workers
```{r, echo = FALSE}
p2
```

To check whether number of socially biased genes is different within and outside the supergene than what would be expected by randomness, the numbers are tested via a Chi squared test. A binomial test is run to check for differences between highly vs lowly expressed genes in polygyne/monogyne individuals. All analyses are performed in queens and workers independently.

```{r enrichment_tests, include = FALSE}
#Workers
to_test_w <- matrix(data = c(de_non_rec_w, non_de_non_rec_w,
                             de_rec_w, non_de_rec_w), nrow = 2)
colnames(to_test_w) <- c("Recombinant", "Non recombinant")
rownames(to_test_w) <- c("Diff expressed", "Non diff expressed")


#Queens
to_test_q <- matrix(data = c(de_non_rec_q, non_de_non_rec_q,
                             de_rec_q, non_de_rec_q), nrow = 2)
colnames(to_test_q) <- c("Recombinant", "Non recombinant")
rownames(to_test_q) <- c("Diff expressed", "Non diff expressed")

```

Chi squared tests for queens:
```{r}
to_test_q
chisq.test(to_test_q)
```

For workers:
```{r}
to_test_w
chisq.test(to_test_w)
```

Binomial tests in queens:
```{r}
#In the genome
binom.test(x = de_rec_m_q, n = de_rec_m_q + de_rec_p_q, p = 0.5)
#In the supergene
binom.test(x = de_non_rec_m_q, n = de_non_rec_m_q + de_non_rec_p_q, p = 0.5)
```

and in workers:
```{r}
#In the genome
binom.test(x = de_rec_m_w, n = de_rec_m_w + de_rec_p_w, p = 0.5)
#In the supergene
binom.test(x = de_non_rec_m_w, n = de_non_rec_m_w + de_non_rec_p_w, p = 0.5)
```
Compare the proportion of monogyne to polygyne biased genes inside and outside the supergene region
```{r compare_mono_poly_bias, include = FALSE}
#Workers
to_test_bias_w <- matrix(data = c(de_rec_m_w, de_rec_p_w,
                                  de_non_rec_m_w, de_non_rec_p_w), nrow = 2)
colnames(to_test_bias_w) <- c("Non Recombinant", "Recombinant")
rownames(to_test_bias_w) <- c("Monogyne biased", "Polygyne biased")


#Queens
to_test_bias_q <- matrix(data = c(de_rec_m_q, de_rec_p_q,
                                  de_non_rec_m_q, de_non_rec_p_q), nrow = 2)
colnames(to_test_bias_q) <- c("Non Recombinant", "Recombinant")
rownames(to_test_bias_q) <- c("Monogyne biased", "Polygyne biased")

```

Chi2 test in queens:
```{r}
to_test_bias_q 
chisq.test(to_test_bias_q)
```
and in workers
```{r}
to_test_bias_w 
chisq.test(to_test_bias_w)
```

## Plot proportion of DE genes by linkage group for queens


```{r, splot_by_lg, include = FALSE}
#This chunk will plot the proportion of DE genes per linkage group, with lg 16 divided into supergene and recombining region.
#First, the two regions of lg need to be separated out as two distinct regions

lg16_regions <- grep(pattern = "lg16", x = loc_genes_region$linkage_group)
loc_genes_region$linkage_group <- as.character(loc_genes_region$linkage_group)
loc_genes_region$linkage_group[lg16_regions] <- paste(loc_genes_region$linkage_group[lg16_regions],
                                                      loc_genes_region$region[lg16_regions], sep = "_")  


#Get the number of DE genes (with M and P bias) in each linkage group
linkage_groups <- unique(loc_genes_region$linkage_group)
loc_genes_region$linkage_group <- as.factor(loc_genes_region$linkage_group)

to_plot_by_lg <- matrix(ncol = 5)
colnames(to_plot_by_lg) <- c("percentage", "region", "form", "nb_genes", "nb_de_genes")
for(lg in linkage_groups){
  #Number of significant genes, M bias
  significant_mono <- length(which(loc_genes_region$linkage_group == lg &
                             loc_genes_region$de_q == "significant" &
                             loc_genes_region$expr_dir_q == "Mono"))
  
  #Number of significant genes, P bias
  significant_poly <- length(which(loc_genes_region$linkage_group == lg &
                             loc_genes_region$de_q == "significant" &
                             loc_genes_region$expr_dir_q == "Poly"))
  
  #Number of non_significant genes
  non_significant <- length(which(loc_genes_region$linkage_group == lg &
                             loc_genes_region$de_q == "non-significant"))
  
  #Total number of genes in each linkage group
  total_nb <- sum(significant_mono, significant_poly, non_significant)
  
  #Total number of DE genes per linkage group
  total_nb_de <- sum(significant_mono, significant_poly)

  #Get proportions
  de_percent_mono <- significant_mono / (non_significant + significant_mono + significant_poly) * 100
  de_percent_poly <- significant_poly / (non_significant + significant_mono + significant_poly) * 100
  
  #Add to data frame
  percents <- c(de_percent_poly, de_percent_mono)
  to_add <- matrix(c(percents,
                     rep(lg, 2),
                     c("poly", "mono"),
                     rep(total_nb, 2),
                     rep(total_nb_de, 2)),
                   ncol = 5)
  to_plot_by_lg <- rbind(to_plot_by_lg, to_add)
}
#Remove NAs fromm the first row
to_plot_by_lg <- to_plot_by_lg[!is.na(to_plot_by_lg[, 1]), ]

to_plot_by_lg <- as.data.frame(to_plot_by_lg)
to_plot_by_lg$percentage <- as.numeric(as.character(to_plot_by_lg$percentage))
to_plot_by_lg$nb_genes <- as.numeric(as.character(to_plot_by_lg$nb_genes))
to_plot_by_lg$nb_de_genes <- as.numeric(as.character(to_plot_by_lg$nb_de_genes))
to_plot_by_lg$region <- factor(to_plot_by_lg$region, levels = unique(to_plot_by_lg$region))

#Barplot
p3 <- ggplot(data = to_plot_by_lg, aes(x = region, y = percentage, fill = form)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = expected_percent_w, colour = "red") +
  theme_bw() + theme(panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
                     legend.title = element_blank()) +
  scale_x_discrete(name = "") + scale_y_continuous(name = "% of differentially expressed genes\nbetween social forms",
                                                   limits = c(0, 8)) +
  scale_fill_manual(values = c("brown4", "cornflowerblue"), labels = c("Single-queen\nbias", "Multiple-queen\nbias")) +
  theme(axis.text.x = element_text(size = 16, angle = 90), axis.text.y = element_text(size = 16),
        axis.title.y = element_text(size = 20))

#Run Chi2 tests for all columns (linkage group vs total proportion of DE vs non-DE)
total_nb_genes <- sum(to_plot_by_lg$nb_genes)/2
total_nb_de_genes <- sum(to_plot_by_lg$nb_de_genes)/2
to_plot_by_lg$pvalue <- NULL
for(lg in linkage_groups){
  lg_to_test <- to_plot_by_lg[to_plot_by_lg$region == lg, ]
  total_genes_lg <- unique(lg_to_test$nb_genes)
  total_genes_de_lg <- unique(lg_to_test$nb_de_genes)
  
  to_test <- matrix(data = c(total_genes_de_lg, (total_genes_lg - total_genes_de_lg),
                             total_nb_de_genes, (total_nb_genes - total_nb_de_genes)), nrow = 2)
  
  p_value <- chisq.test(to_test)$p.value
  to_plot_by_lg$pvalue[to_plot_by_lg$region == lg] <- p_value
  }

to_plot_by_lg$padj <- p.adjust(to_plot_by_lg$pvalue, method = "BH")
```

Run a quasi-binomial model to check the enrichment of the supergene compared to the rest of linkage groups

```{r, glm_enrichment, include = FALSE}

#Extract number of differentially expressed genes as a matrix for the dependent variable of the glm
de_genes_ratio <- matrix(c(to_plot_by_lg$nb_de_genes, to_plot_by_lg$nb_genes), ncol = 2)
de_genes_ratio <- unique(de_genes_ratio)
colnames(de_genes_ratio) <- c("de_genes", "all_genes")

#Extract the predicting variables: lg names and a binary variable for supergene/non-supergene
supergene_bin <- as.factor(ifelse(linkage_groups == "lg16_supergene", 2, 1))
linkage_groups <- as.factor(linkage_groups)

#Run the model
supergene_enrichment <- glm(de_genes_ratio ~ supergene_bin, family = "quasibinomial")

#Significance levels
summary(supergene_enrichment)
#Run a similar model for all linkage groups (+ p value correction)
#Remove the superge from the data frame and the list of lg
de_genes_ratio_no_super <- de_genes_ratio[!linkage_groups == "lg16_supergene", ]
linkage_groups_no_super <- linkage_groups[!linkage_groups == "lg16_supergene"]

#Generate a dataframe for the new pvalues
glm_non_super_lg <- data.frame(linkage_groups_no_super)
glm_non_super_lg$pvalue <- NULL

colnames(glm_non_super_lg)[1] <- "linkage_group"

for(lg in linkage_groups_no_super){
  
  lg_bin <- as.factor(ifelse(linkage_groups_no_super == lg, 2, 1))
  lg_enrichment <- glm(de_genes_ratio_no_super ~ lg_bin, family = "quasibinomial")
  
  raw_pval <- summary(lg_enrichment)$coefficients[2, 4]
  glm_non_super_lg$pvalue[glm_non_super_lg$linkage_group == lg] <- raw_pval
}

#Correct pvalues 
glm_non_super_lg$padj <- p.adjust(glm_non_super_lg$pvalue, method = "BH")

#Which other linkage groups are enriched?
glm_non_super_lg[glm_non_super_lg$padj < 0.05, ]

```


## Plot enrichment and differences poly/mono separately

Plot the enrichment of the supergene compared to the rest of the genome and 

```{r, plot_enrichment_social_forms_diffs, include = FALSE}

to_plot_q$region <- factor(to_plot_q$region, levels = c("Supergene",  "Rest of the genome"))

penrich <- ggplot(data = to_plot_q, aes(x = region, y = percentage)) +
              geom_bar(stat = "identity", fill = "steelblue") +
              theme_bw() + theme(panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
                                 legend.title = element_blank()) +
              scale_x_discrete(name = "", labels = c("Supergene", "Rest of\nthe genome")) + scale_y_continuous(name = "% differentially expressed genes",
                                                               limits = c(0, 9)) +
              theme(axis.text.x = element_text(size = 24), axis.text.y = element_text(size = 24),
                    axis.title.y = element_text(size = 28), legend.text = element_text(size = 24))

psoc_form <- ggplot(data = to_plot_q, aes(x = form, y = percentage)) +
              geom_bar(stat = "identity", fill = "steelblue") +
              theme_bw() + theme(panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
                                 legend.title = element_blank()) +
              scale_x_discrete(name = "", labels = c("Supergene", "Rest of\nthe genome")) + scale_y_continuous(name = "% differentially expressed genes",
                                                               limits = c(0, 9)) +
              theme(axis.text.x = element_text(size = 24), axis.text.y = element_text(size = 24),
                    axis.title.y = element_text(size = 28), legend.text = element_text(size = 24))

#Plot polygine bias within and outside the supergene region

tot_percent_de_nrec <- de_percent_nrec_p_q + de_percent_nrec_m_q
tot_percent_de_rec <- de_percent_rec_p_q + de_percent_rec_m_q

poly_bias_nrec <- (de_percent_nrec_p_q / tot_percent_de_nrec) * 100
poly_bias_rec <- (de_percent_rec_p_q / tot_percent_de_rec) * 100

to_plot_poy_bias <- data.frame(c(poly_bias_nrec, poly_bias_rec),
                               factor(c("Supergene",  "Rest of the genome"),
                                      levels = c("Supergene",  "Rest of the genome")))

colnames(to_plot_poy_bias ) <- c("poly_bias", "region")

ppoly_bias <- ggplot(data = to_plot_poy_bias, aes(x = region, y = poly_bias)) +
              geom_bar(stat = "identity", fill = "brown4") +
              theme_bw() + theme(panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
                                 legend.title = element_blank()) +
              scale_x_discrete(name = "", labels = c("Supergene", "Rest of\nthe genome")) + scale_y_continuous(name = "% differentially expressed genes") +
              theme(axis.text.x = element_text(size = 24), axis.text.y = element_text(size = 24),
                    axis.title.y = element_text(size = 28), legend.text = element_text(size = 24))


ggsave(filename = "results/enrichment_supergene.pdf", plot = penrich, device = "pdf", height = 20, width = 15, units = "cm")
ggsave(filename = "results/social_form_bias.pdf", plot = psoc_form, device = "pdf", height = 20, width = 15, units = "cm")
ggsave(filename = "results/polygyne_bias_by_region.pdf", plot = ppoly_bias, device = "pdf", height = 20, width = 15, units = "cm")
```

## Supplementary tables

```{r, supp_tables, include = FALSE}
#Generate results table only with significant genes for each comparison
res_de_q_sig <- res_de_q[which(res_de_q$padj < 0.05), ]
res_de_w_sig <- res_de_w[which(res_de_w$padj < 0.05), ]

df_de_q_sig <- as.data.frame(res_de_q_sig)
df_de_w_sig <- as.data.frame(res_de_w_sig)

```






