---
title: "Sb to Sb differences heatamp"
author: "Carlos Martinez Ruiz"
date: "18 January 2019"
output: pdf_document
---

```{r setup, include = FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
```{r libraries, echo = FALSE, message = FALSE}
#Load all libraries
library(tximport)
library(readr)
library(DESeq2)
library(GenomicRanges)
library(GenomicFeatures)
library(ggplot2)
library(plyr)
version

for (package in (.packages()) ) {
 print(paste("Package", package, "version", package.version(package)))
}
```

Load the data from the Sb vs SB allele specific comparison for both North American and South American populations, extract read counts. 

```{r load_data, include = FALSE}
load("input/dds_Sb_vs_SB_north_america.RData")
load("input/dds_Sb_vs_SB_south_america.RData")

counts_north_america <- counts(dds_Bb_DE)
counts_south_america <- counts(dds_deg_ar)
``` 

How many genes do both datasets have in common?

```{r genes_in_common, include = FALSE}
genes_south_america <- row.names(counts_south_america)
genes_north_america <- row.names(counts_north_america)

genes_both <- genes_north_america[genes_north_america %in% genes_south_america]

#Number of genes present in both datsets:
length(genes_both)
``` 

All genes found with fixed differences in South America are also present in North America

##Plot heatmap

Heatmap with raw read counts
```{r plot_heatmap_read_counts, include = FALSE}
#Parse the dataset to make it ggplot friendly
samples_north_america <- colnames(counts_north_america)
parsed_counts_north_america <- data.frame(c(counts_north_america),
                                          rep(samples_north_america, each = nrow(counts_north_america)),
                                          rep(genes_north_america, ncol(counts_north_america)), 
                                          rep("north_america", (nrow(counts_north_america) * ncol(counts_north_america)))
                                          )
colnames(parsed_counts_north_america) <- c("counts", "sample", "gene", "population") 

samples_south_america <- colnames(counts_south_america)
parsed_counts_south_america <- data.frame(c(counts_south_america),
                                          rep(samples_south_america, each = nrow(counts_south_america)),
                                          rep(genes_south_america, ncol(counts_south_america)), 
                                          rep("south_america", (nrow(counts_south_america) * ncol(counts_south_america)))
                                          )
colnames(parsed_counts_south_america) <- c("counts", "sample", "gene", "population") 

#Add the missing genes in the south american dataset to make both datasets compatible
missing_genes_south_america <- genes_north_america[!genes_north_america %in% genes_south_america]
add_missing_genes <- data.frame(rep(NA, length(missing_genes_south_america) * ncol(counts_south_america)),
                                rep(samples_south_america, length(missing_genes_south_america)),
                                rep(missing_genes_south_america, ncol(counts_south_america)),
                                rep("south_america", length(missing_genes_south_america) * ncol(counts_south_america))
                                )
colnames(add_missing_genes) <- c("counts", "sample", "gene", "population") 

parsed_counts_south_america <- rbind(parsed_counts_south_america, add_missing_genes)
rm(add_missing_genes)

#Merge both datasets
parsed_counts_all <- rbind(parsed_counts_north_america, parsed_counts_south_america)
parsed_counts_all$allele <- gsub(x = parsed_counts_all$sample, pattern = ".+_", replacement = "")
parsed_counts_all$allele <- ifelse(parsed_counts_all$allele == "B" , "bigB", "littleb")
parsed_counts_all <- parsed_counts_all[order(parsed_counts_all$allele, parsed_counts_all$sample), ]

#Plot heatmap
ggplot(parsed_counts_all, aes(x = sample, y = gene)) + geom_tile(aes(fill = log2(counts + 1)), colour = "grey50") +
  facet_grid(. ~ population, scales = "free", space = "free") +
  scale_fill_gradient(low = "yellow" , high = "blue", name="log2 Estimated \nnormalized counts") + 
  theme_bw() + theme( panel.grid.major=element_blank()) + theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
  theme(axis.title.x = element_blank(), axis.title.y = element_blank()) + 
  theme(strip.text.x = element_text(size=20)) + theme(legend.title = element_text(size=15, face="bold"),
                                                      legend.text = element_text(size=13, face="bold")) 

``` 


Heatmap with logarithm of the read count ratios between variants for North American populations
```{r plot_heatmap_read_counts_north_america, include = FALSE}
#Parse the dataset to make it ggplot friendly
#Separate big B and little b samples
samples_north_america_littleb <- grep(x = samples_north_america, pattern = "_b", value = TRUE)
samples_north_america_bigB <- grep(x = samples_north_america, pattern = "_B", value = TRUE)
samples_north_america_no_allele <- unique(gsub(x = samples_north_america, pattern = "_[Bb]", replacement = ""))

counts_north_america_littleb <- counts_north_america[, samples_north_america_littleb]
counts_north_america_bigB <- counts_north_america[, samples_north_america_bigB]


parsed_counts_north_america_ratio <- data.frame(c(counts_north_america_littleb),
                                                c(counts_north_america_bigB),
                                                rep(samples_north_america_no_allele, each = nrow(counts_north_america)),
                                                rep(genes_north_america, length(samples_north_america_no_allele)), 
                                                rep("north_america", (nrow(counts_north_america) * length(samples_north_america_no_allele)))
                                                )
colnames(parsed_counts_north_america_ratio) <- c("littleb", "bigB", "sample", "gene", "population") 

parsed_counts_north_america_ratio$total_counts <- parsed_counts_north_america_ratio$littleb + parsed_counts_north_america_ratio$bigB
parsed_counts_north_america_ratio$lfc <- log2((parsed_counts_north_america_ratio$bigB +1) / (parsed_counts_north_america_ratio$littleb +1))

#Order the genes by mean read count
mean_reads_north_america <- ddply(parsed_counts_north_america_ratio, .(gene), summarize, mean = 
                                  mean(total_counts))
mean_reads_north_america <- mean_reads_north_america[order(mean_reads_north_america$mean), ]
gene_order_north_america <- rep(mean_reads_north_america$gene,
                                length(unique(samples_north_america_no_allele)))

#Use the sorted genes to sort the parsed counts
parsed_counts_north_america_ratio$gene <- factor(parsed_counts_north_america_ratio$gene, levels = mean_reads_north_america$gene)
  
#Plot heatmap for North America
ggplot(parsed_counts_north_america_ratio, aes(x = sample, y = gene)) + geom_tile(aes(fill = lfc), colour = "grey50") +
       scale_fill_gradient(low = "yellow" , high = "blue", name="log2 Estimated \nnormalized counts") + 
       theme_bw() + theme( panel.grid.major=element_blank()) + theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
       theme(axis.title.x = element_blank(), axis.title.y = element_blank()) + 
       theme(strip.text.x = element_text(size=20)) + theme(legend.title = element_text(size=15, face="bold"),
                                                           legend.text = element_text(size=13, face="bold")) 



```

Comparison of mean read counts between South American and North American populations
```{r compare_mean_read_counts, include = FALSE}
#Parse the dataset to make it ggplot friendly
#Separate big B and little b samples
samples_south_america_littleb <- grep(x = samples_south_america, pattern = "_b", value = TRUE)
samples_south_america_bigB <- grep(x = samples_south_america, pattern = "_B", value = TRUE)
samples_south_america_no_allele <- unique(gsub(x = samples_south_america, pattern = "_[Bb]", replacement = ""))

counts_south_america_littleb <- counts_south_america[, samples_south_america_littleb]
counts_south_america_bigB <- counts_south_america[, samples_south_america_bigB]

parsed_counts_south_america_ratio <- data.frame(c(counts_south_america_littleb),
                                                c(counts_south_america_bigB),
                                                rep(samples_south_america_no_allele,
                                                    each =nrow(counts_south_america)),
                                                rep(genes_south_america,
                                                    length(samples_south_america_no_allele)), 
                                                rep("south_america",
                                                    (nrow(counts_south_america) * length(samples_south_america_no_allele)))
                                                )
colnames(parsed_counts_south_america_ratio) <- c("littleb", "bigB", "sample", "gene", "population") 

parsed_counts_south_america_ratio$total_counts <- parsed_counts_south_america_ratio$littleb + parsed_counts_south_america_ratio$bigB
parsed_counts_south_america_ratio$lfc <- log2((parsed_counts_south_america_ratio$bigB +1) / (parsed_counts_south_america_ratio$littleb +1))

#Get mean read count for South American samples
mean_reads_south_america <- ddply(parsed_counts_south_america_ratio, .(gene), summarize, mean = 
                                  mean(total_counts))
mean_reads_south_america <- mean_reads_south_america[order(mean_reads_south_america$mean), ]

#Get the list of genes in North America that overlap with South America
mean_reads_all <- merge(mean_reads_north_america, mean_reads_south_america, by = "gene")
colnames(mean_reads_all) <- c("gene", "mean_na", "mean_sa")

#Plot means per population
plot(log(mean_reads_all$mean_na),
     log(mean_reads_all$mean_sa))

```


