---
title: "ase_comparison_south_america.Rmd"
author: "Carlos Martinez Ruiz"
date: "20 de noviembre de 2018"
output: pdf_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "C:/Users/Carlos Martinez Ruiz/Dropbox/PhD")
``` 
```{r libraries, echo = FALSE, message = FALSE}
#Load all libraries
library(tximport)
library(readr)
library(DESeq2)
library(GenomicRanges)
library(GenomicFeatures)
library(ggplot2)

for (package in (.packages()) ) { 
 print(paste("Package", package, "version", package.version(package)))
}
```

## Introduction

This script will analyse the variant specific expression for the supergene South American populations from *Solenopsis invicta*. The RNAseq data used for this analysis was extracted from three different body parts (abdomen, thorax and head) from virgin alate queens and whole body from workers, all from polygyne colonies. All individuals were genotyped prior to RNA extraction to make sure that they are heterozygote Sb/SB for the social supergene. To obtain the variant-specific reads, the RNAseq reads were aligned to the *S.invicta* reference genome (gnG assembly) and the variant specific read counts obtained by running GATK's ASEreadCounter with a VCF file specific for North American populations. 

## Import variant specific reads into R

The read counts generated by GATK are stored in a csv file, where one of the columns has the read counts for the reference allele (SB) and another one for the alternative (Sb).

```{r load_data_r, message = FALSE}
#Loads the counts by variant generated by GATK. All samples!!
ase_files_ar <- grep("csv", list.files("DC_sinvicta_2016-17/11_2016_ase_analysis/2018-10-13_ar_subset_ncoding/"), value = TRUE)
ase_files_ar <- paste("DC_sinvicta_2016-17/11_2016_ase_analysis/2018-10-13_ar_subset_ncoding/", ase_files_ar, sep="" )

#Add names to the samples
names(ase_files_ar) <- gsub(x = ase_files_ar, pattern = "(.+/)([0-9]+[A-Z]+)(_.+)", "\\2", perl = TRUE)

#Remove samples from colony AR117, which had very low mapping scores
ase_files_ar <- ase_files_ar[-grep(x = ase_files_ar, pattern = "117")]
#ase_files_ar <- ase_files_ar[-grep(x = ase_files_ar, pattern = "28")]
#ase_files_ar <- ase_files_ar[-grep(x = ase_files_ar, pattern = "104")]
ase_files_ar

#Load all the files in R as a single table
ase_ar <- do.call(rbind,lapply(ase_files_ar, read.table, header = TRUE)) 

#The table loaded contains all ase counts for the AR colonies. This includes different castes (W and Q) and different tissues within queens (Head, Thorax and Abdomen). 
#Each file has different lengths since not all individuals have the same variants.

#Transforms the table loaded into a GRanges object, readable by GenomicRanges and GenomicFeatures.
#The first two parameters define the chromosome (contig in this case) and the position of the feature (in this case,
# variants). The other parameters are metadata (they could be whatever) in this case, the counts for the
# reference, the alternative allele and total and the sample to which each read belongs.

ase_gr <- GRanges(Rle(ase_ar$contig), IRanges(start = ase_ar$position, width = 1), countRef = ase_ar$refCount, countAlt = ase_ar$altCount,
                  totCount = ase_ar$totalCount, sample = gsub("([0-9]+[A-Z]+)(\\.[0-9]+)", "\\1", rownames(ase_ar)))
```