---
title: "dosage_compensation_plot"
author: "Carlos Martinez Ruiz"
date: "15 June 2020"
output: pdf_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.path = 'results/')
```
```{r libraries, echo = FALSE, message = FALSE}
#Load all libraries

library(DESeq2)
library(ggplot2)
library(edgeR)
library(rcompanion)
library(dplyr)

version

for (package in (.packages()) ) {
 print(paste("Package", package, "version", package.version(package)))
}
```
Make a similar plot to Fig.3 in 'Rapid De Novo Evolution of X Chromosome Dosage Compensation in Silene latifolia, a Plant with Young Sex Chromosomes' (PLoS Biology, 2012). Basically compare the expression ratio of SB/Sb against expression of the same genes in SBSB and SBSb individuals.

Load the data needed for the plot: LFCs between SB and Sb for North America generated by DESeq2 and average expression for the same genes in monogyne and polygyne.

```{r load_data_r, message = FALSE}
#Load lfcs for both comparisons 
load("results/lfcs_joint.RData")
#Select genes to exclude: Either high expression in SB and Polygynes or high Sb expression
to_exclude <- lfcs_joint_nona$gene[(lfcs_joint_nona$padj_ase < 0.05 & lfcs_joint_nona$lfcs_ase < 0) |
                                   (lfcs_joint_nona$padj_ase < 0.05 & lfcs_joint_nona$lfcs_ase > 0 &
                                    lfcs_joint_nona$padj_mor < 0.05 & lfcs_joint_nona$lfcs_mor > 0)]

to_exclude <- as.character(to_exclude)
#Load normalised expression data for social forms
load("input/2020-03-01_dds_deseq2_queens_mor.RData")
soc_form_norm <- counts(de_txi_q, normalized = TRUE)
soc_form_res <- results(de_txi_q)

#Load LFCs between SB and Sb and counts (normalised really does nothing because ASE reads were normalised by 1)
load("results/dds_Sb_vs_SB_north_america_corr.RData")
res_ase <- results(dds_Bb_DE)
ase_norm <- counts(dds_Bb_DE, normalized = TRUE)

```

Once the data is loaded, plot median Mono (SBSB) and Polygyne (SBSb) expression and PolySb and PolySB for all genes with LFC information between variants

```{r plot_dosage_compensation, message = FALSE}
#Extract only genes in common in both comparisons
genes_common <- rownames(soc_form_norm)[rownames(soc_form_norm) %in% rownames(res_ase)]

#Subset to keep only common genes
soc_form_common <- soc_form_norm[genes_common, ]
res_ase_common <- res_ase[genes_common, ]
ase_norm_common <- ase_norm[genes_common, ]

#Transform lfcs so that high lfc implies high Sb expression
res_ase_common$log2FoldChange <- -(res_ase_common$log2FoldChange)
#Transform lfcs into ratios
res_ase_common$ratios <- 2^res_ase_common$log2FoldChange
#Get the ase ratios between variants per gene
dos_comp_df <- data.frame(gene = rownames(res_ase_common), ase_ratio = res_ase_common$ratios)

#Generate a dataframe with the read counts per gene for all four categories: SBSB, SBSb, PolySB and PolySB
#(17 replicates in total, 3 + 2 + 6 + 6 respectively)
dos_comp_df <- data.frame(gene = rep(rownames(res_ase_common), 17),
                          ase_ratio = rep(res_ase_common$ratios, 17),
                          expr_levels = c(c(soc_form_common[, grep("Mono", colnames(soc_form_common))]),
                                        c(soc_form_common[, grep("Poly", colnames(soc_form_common))]),
                                        c(ase_norm_common[, grep("_B", colnames(ase_norm_common))]),
                                        c(ase_norm_common[, grep("_b", colnames(ase_norm_common))])),
                          expr_group = c(rep("Mono", nrow(soc_form_common)*3), rep("Poly", nrow(soc_form_common)*2),
                                         rep("PolySB", nrow(ase_norm_common)*6), rep("PolySb", nrow(ase_norm_common)*6)),
                          stringsAsFactors = FALSE)

#Remove genes with ither high expression in SB and Polygynes or high Sb expression
dos_comp_df_exc <- dos_comp_df[!dos_comp_df$gene %in% to_exclude, ]

#Generate categories for ratios
dos_comp_df_exc$ratios_cat <- NA

dos_comp_df_exc$ratios_cat[dos_comp_df_exc$ase_ratio > 0 & dos_comp_df_exc$ase_ratio <= 0.5] <- "0-0.5"
dos_comp_df_exc$ratios_cat[dos_comp_df_exc$ase_ratio > 0.5 & dos_comp_df_exc$ase_ratio <= 0.75] <- "0.5-0.75"
dos_comp_df_exc$ratios_cat[dos_comp_df_exc$ase_ratio > 0.75 & dos_comp_df_exc$ase_ratio <= 1] <- "0.75-1"
dos_comp_df_exc$ratios_cat[dos_comp_df_exc$ase_ratio > 1 & dos_comp_df_exc$ase_ratio <= 1.5] <- "1-1.5"
dos_comp_df_exc$ratios_cat[dos_comp_df_exc$ase_ratio > 1.5] <- ">1.5"

#Divide each expression group by its median to make them comparable
medians_expr_group <- groupwiseMedian(expr_levels ~ expr_group, data = dos_comp_df_exc, boot = FALSE, bca = FALSE)
dos_comp_df_exc$expr_levels_norm <- NA

for(this_group in unique(dos_comp_df_exc$expr_group)){
  this_median <- medians_expr_group$Median[medians_expr_group$expr_group == this_group]
  dos_comp_df_exc$expr_levels_norm[dos_comp_df_exc$expr_group == this_group] <-
        dos_comp_df_exc$expr_levels[dos_comp_df_exc$expr_group == this_group]/this_median
}

#Get medians and 95% CI (from means of bootstrapped medians) to plot the thing
dos_comp_df_exc_to_plot <- groupwiseMedian(expr_levels_norm ~ expr_group + ratios_cat, data = dos_comp_df_exc,
                                           conf = 0.95,
                                           R = 5000,
                                           wilcox = TRUE,
                                           percentile = TRUE,
                                           bca = FALSE,
                                           digits = 3)

#Reorder the different factors
dos_comp_df_exc_to_plot$ratios_cat <- factor(dos_comp_df_exc_to_plot$ratios_cat,
                                             levels = c("0-0.5", "0.5-0.75", "0.75-1", "1-1.5", ">1.5"))

dos_comp_df_exc_to_plot$expr_group <- factor(dos_comp_df_exc_to_plot$expr_group,
                                             levels = c("Mono", "PolySB", "PolySb", "Poly"))

barplot_colors <- c("#e66101", "#fdb863", "#b2abd2", "#5e3c99")

#Plot the thing
dos_comp_plot <- ggplot(dos_comp_df_exc_to_plot, aes(x = ratios_cat, y = Median, fill = expr_group)) +
                    geom_bar(stat = "identity", position=position_dodge()) +
                    geom_errorbar(aes(ymin = Percentile.lower, ymax = Percentile.upper),
                                  width=.2, position=position_dodge(.9)) +
                    labs(x = "Sb/SB expression ratio", y = "Median normalised counts") +
                    scale_fill_manual(values = barplot_colors, name = "", labels = c("Single queen (SBSB)", "Multiple queen SB",
                                                                                     "Multiple queen Sb", "Multiple queen (SBSb)")) +
                    theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                       panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                       axis.title.x = element_text(size = 20),
                                       axis.text.x  = element_text(vjust = 0.5, size = 18),
                                       axis.title.y = element_text(size = 20),
                                       axis.text.y  = element_text(vjust = 0.5, size = 18),
                                       legend.title = element_text(size = 18),
                                       legend.text = element_text(size = 15))

#Save plot
ggsave(dos_comp_plot, filename = "results/dosage_compensation_plot.pdf", device = "pdf",
       width = 29.7, height = 21, units = "cm")

wilcox_pvals_cats <- matrix(NA, ncol = 2, dimnames = list(NULL, c("category", "p_value")))
#Calculate significance differences in median between Mono and Poly in each allele-specific ratio category
for(this_cat in unique(dos_comp_df_exc$ratios_cat)){
  subset_cat <- dos_comp_df_exc[dos_comp_df_exc$ratios_cat == this_cat, ]
  mono_cat <- subset_cat$expr_levels_norm[subset_cat$expr_group == "Mono"]
  poly_cat <- subset_cat$expr_levels_norm[subset_cat$expr_group == "Poly"]
  wilcox_results <-  wilcox.test(mono_cat, poly_cat)
  to_add <- c(this_cat, wilcox_results$p.value)
  wilcox_pvals_cats <- rbind(wilcox_pvals_cats, to_add)
}
#None of the comparisons are significant

#Try using mean instead of median and sd instead of CI
dos_comp_df_exc %>% group_by(ratios_cat, expr_group) %>% summarise(mean = mean(expr_levels_norm), sd = sd(expr_levels_norm))

```

