---
title: "ase_analyisis_south_america_whole_genome.Rmd"
author: "Carlos Martinez Ruiz"
date: "5 July 2019"
output: pdf_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
``` 
```{r libraries, echo = FALSE, message = FALSE}
#Load all libraries
library(tximport)
library(readr)
library(DESeq2)
library(GenomicRanges)
library(GenomicFeatures)
library(ggplot2)
library(dplyr)

for (package in (.packages()) ) { 
 print(paste("Package", package, "version", package.version(package)))
}
```

## Introduction

This script will analyse the variant specific expression for the supergene South American populations from *Solenopsis invicta*. The RNAseq data used for this analysis was extracted from three different body parts (abdomen, thorax and head) from virgin alate queens and whole body from workers, all from polygyne colonies. All individuals were genotyped prior to RNA extraction to make sure that they are heterozygote Sb/SB for the social supergene. To obtain the variant-specific reads, the RNAseq reads were aligned to the *S.invicta* reference genome (gnG assembly) and the variant specific read counts obtained by running GATK's ASEreadCounter with a VCF file specific for North American populations. In this particular case, the aim is to assess whether we can detect ASE differences between body parts using ASE from the whole genome (previous analyses were performed on the supergene region only). Because we are now using data from the whole genome (with no phase information) we cannot group the SNPs per gene, as they could belong to different haplotypes. We also need to filter out uninformative SNPs where there is no expression in one of the alleles (likely to be a fixed a llele)

## Import variant specific reads into 

The read counts generated by GATK are stored in a csv file, where one of the columns has the read counts for the reference allele (SB) and another one for the alternative (Sb).

```{r load_data_r, message = FALSE}
#Loads the counts by variant generated by GATK. All samples!!
path_to_samples <- "input/"
ase_files_ar <- grep("csv", list.files(path_to_samples), value = TRUE)
ase_files_ar <- paste(path_to_samples, ase_files_ar, sep="" )

#Add names to the samples
names(ase_files_ar) <- gsub(x = ase_files_ar, pattern = "(.+/)([0-9]+[A-Z]+)(_.+)", "\\2", perl = TRUE)

#Remove samples from colony AR117, which had very low mapping scores
ase_files_ar <- ase_files_ar[-grep(x = ase_files_ar, pattern = "117")]
#ase_files_ar <- ase_files_ar[-grep(x = ase_files_ar, pattern = "28")]
#ase_files_ar <- ase_files_ar[-grep(x = ase_files_ar, pattern = "104")]
ase_files_ar

#Load all the files in R as a single table
ase_ar <- do.call(rbind,lapply(ase_files_ar, read.table, header = TRUE, stringsAsFactors = FALSE)) 

#The table loaded contains all ase counts for the AR colonies. This includes different castes (W and Q) and different tissues within queens (Head, Thorax and Abdomen). 
#Each file has different lengths since not all individuals have the same variants.
```

## Load variant specific read counts into R

Parse the dataset to make it compatible for DESeq2 and then filter out SNPs with very few reads (less than 5 in total) and with low reads in one of the alleles only (less than 3 for any allele).
```{r parse_ase_data, message = FALSE}

#Generate new columns with SNP ids and the sample name
ase_ar$snp_id <- paste(ase_ar$contig, ase_ar$position, sep = "_")
ase_ar$sample <- gsub(pattern = "\\.[0-9]+",
                               replacement = "",
                               x = rownames(ase_ar))

#Keep only the columns of interest
ase_ar_parsed <- data.frame(sample = ase_ar$sample, snp_id = ase_ar$snp_id,
                            ref = ase_ar$refCount, alt = ase_ar$altCount)

#Keep only SNPs which are present in all samples
ids_to_keep <- names(which(table(ase_ar_parsed$snp_id) == length(unique(ase_ar_parsed$sample))))
ase_ar_parsed <- ase_ar_parsed[ase_ar_parsed$snp_id %in% ids_to_keep, ]

#Generate a matrix for DESeq2 with the snp ids in the rownames and the sample name in the columns
#For reference alleles first
ase_mat_ref <- matrix(c(ase_ar_parsed$ref),
                      nrow = length(unique(ase_ar_parsed$snp_id)),
                      ncol = length(unique(ase_ar_parsed$sample)),
                      dimnames = list(unique(ase_ar_parsed$snp_id), 
                                      paste0(unique(ase_ar_parsed$sample), "_ref")))

#Get the names of snps with an average read count lower than 3
snps_to_filter_ref <- names(which(apply(ase_mat_ref, 1, mean) < 3))

#For alternative alleles
ase_mat_alt <- matrix(c(ase_ar_parsed$alt),
                      nrow = length(unique(ase_ar_parsed$snp_id)),
                      ncol = length(unique(ase_ar_parsed$sample)),
                      dimnames = list(unique(ase_ar_parsed$snp_id), 
                                      paste0(unique(ase_ar_parsed$sample), "_alt")))

#Get the names of snps with an average read count lower than 3
snps_to_filter_alt <- names(which(apply(ase_mat_alt, 1, mean) < 3))

#Get the amtrix for all counts
ase_mat <- cbind(ase_mat_ref, ase_mat_alt)

#Get a vector of all the snps to remove for low counts and remove them from the main matrix
snps_to_filter <- unique(c(snps_to_filter_alt, snps_to_filter_ref))

ase_mat <- ase_mat[which(!rownames(ase_mat) %in% snps_to_filter), ]

#Generate metadata for DESeq2
whole_sample    <- gsub(x = colnames(ase_mat), pattern = "_.+", replacement = "")
allele          <- gsub(x = colnames(ase_mat), pattern = ".+_", replacement = "")
body_part       <- gsub(x = whole_sample, pattern = "([0-9]+[BC]?Q?)([A-Z])", replacement = "\\2")
colony <- gsub(x = whole_sample, pattern = "([0-9]+)([A-Z]+)",
               replacement = paste0("AR", "\\1"),
               perl = TRUE)
colData_ar <- data.frame(whole_sample, allele, body_part, colony)
rownames(colData_ar) <- colnames(ase_mat)
```

Import data into DESeq2

##DESeq2 analysis

The analysis for the allele specific expression between the alternative and reference alleles is performed in DESeq2. Because the alelle specific counts that are being compared always come from the same sample (and therefore, the same library), performing normalisation would flatten actual differences. The analysis therefore needs to be performed ensuring that the samples are not normalised (more info here<http://rpubs.com/mikelove/ase>). The first analysis will consider 'colony', 'body_part', 'allele' and the interaction between 'allele:body_part'.

``` {r interaction_allele:body_part, message = FALSE}
#Get the data into DESeq2, get size factors by body part only (or sample?)
dds_ar_interaction <- DESeqDataSetFromMatrix(countData = ase_mat, colData = colData_ar, design = ~ colony + body_part + allele + allele:body_part)
#Perform the analysis replacing the sizeFactors: 
sizeFactors(dds_ar_interaction) <- rep(1, ncol(ase_mat))
#If using test = LRT, deseq2 performs a test for detecting DE loci, using first a likelihood ratio test (ANOVA-like). This test compares the model ~body_part+allele against a
#reduced model, ~body_part. The p values indicate which genes are significantly DE accross ALL levels of allele. Normalisation
#has been done only by boy_part, as ASE data comes always from the same sample. If test is not specified it runs the usual Wald test for the last element in the design formula (body_part:allele)
dds_deg_ar_interaction <- DESeq(dds_ar_interaction, test = "LRT", reduced = ~ colony + body_part + allele, betaPrior = FALSE)

#Get results
res_deg_ar_interaction <- results(dds_deg_ar_interaction)
#How many DEG with interaction?
length(which(res_deg_ar_interaction$padj < 0.05))
#18
```
Out of the 2599 genes analysed, 7 show allele specific expression differences between body parts.

##Variant specific vs body part comparisons----------------CONTINUE HERE-------------------------

Different body parts will perform different functions, and this is reflected in gene expression differences. If most genes in the supergene region perform different functions related to social form, the expectation would be that the SB to Sb differences across body parts should change, as we would expect tissues in the head to be performing differently to those in the abdomen. We already know that there is no Sb-Sb, body part interaction from the previous test, but plotting the SB-Sb differences across tissues will help visualise this result.

``` {r plot_allele:body_part_interaction, message = FALSE}
#Get coefficients for interactions only
dds_ar_coef <- DESeqDataSetFromMatrix(countData = to_analyse_DESeq2_ar, colData = colData_ar, design = ~ colony + body_part + allele:body_part)
#Perform the analysis replacing the sizeFactors: 
sizeFactors(dds_ar_coef) <- rep(1, ncol(to_analyse_DESeq2_ar))
dds_deg_ar_coef <- DESeq(dds_ar_coef, betaPrior = FALSE)
res_deg_ar_coef <- results(dds_deg_ar_coef)

coef_QA <- coefficients(dds_deg_ar_coef)[, "body_partA.alleleB"]
coef_QT <- coefficients(dds_deg_ar_coef)[, "body_partT.alleleB"]
coef_QH <- coefficients(dds_deg_ar_coef)[, "body_partH.alleleB"]
coef_W  <- coefficients(dds_deg_ar_coef)[, "body_partW.alleleB"]

#Plot weihting points by read count mean, add line with linear regression, weighted by counts
#Weight for the point size, by log(mean counts per gene), scaled up to fit 0-1 range
cex_weights <- (log(res_deg_ar_coef$baseMean) - min(log(res_deg_ar_coef$baseMean)))/
  (max(log(res_deg_ar_coef$baseMean))-min(log(res_deg_ar_coef$baseMean)))
par(mfrow = c(3, 2), cex.lab = 2, cex.axis = 2)
plot(coef_QA, coef_QT, type = "n", xlab = "Queen Abdomen LFCs", ylab = "Queen Thorax LFCs")
points(coef_QA, coef_QT, cex = 2*cex_weights, pch = 20)
abline(0,1)
abline(lm(coef_QT ~ coef_QA, weights = res_deg_ar_coef$baseMean), col = "red")

plot(coef_QA, coef_QH, type = "n", xlab = "Queen Abdomen LFCs", ylab = "Queen Head LFCs")
points(coef_QA, coef_QH, cex = 2*cex_weights, pch = 20)
abline(0,1)
abline(lm(coef_QH ~ coef_QA, weights = res_deg_ar_coef$baseMean), col = "red")

plot(coef_QA, coef_W, type = "n", xlab = "Queen Abdomen LFCs", ylab = "Worker LFCs")
points(coef_QA, coef_W, cex = 2*cex_weights, pch = 20)
abline(0,1)
abline(lm(coef_W ~ coef_QA, weights = res_deg_ar_coef$baseMean), col = "red")

plot(coef_QH, coef_W, type = "n", xlab = "Queen Head LFCs", ylab = "Worker LFCs")
points(coef_QH, coef_W, cex = 2*cex_weights, pch = 20)
abline(0,1)
abline(lm(coef_W ~ coef_QH, weights = res_deg_ar_coef$baseMean), col = "red")

plot(coef_QT, coef_W, type = "n", xlab = "Queen Thorax LFCs", ylab = "Worker LFCs")
points(coef_QT, coef_W, cex = 2*cex_weights, pch = 20)
abline(0,1)
abline(lm(coef_W ~ coef_QT, weights = res_deg_ar_coef$baseMean), col = "red")

plot(coef_QT, coef_QH, type = "n", xlab = "Queen Thorax LFCs", ylab = "Queen Head LFCs")
points(coef_QT, coef_QA, cex = 2*cex_weights, pch = 20)
abline(0,1)
abline(lm(coef_QA ~ coef_QT, weights = res_deg_ar_coef$baseMean), col = "red")
par(mfrow = c(1,1))
```

The LRT showed no significant interaction terms in general. Here we run a pairwise Wald test comparing each interaction to the reference level (alleleb, i.e. the effect of the Sb variant) to double check that no interactions are significant
``` {r pairwise_comparisons_allele:body_part_interaction, message = FALSE}
#Extract all the names for potential contrasts, keep only the interactions
dds_deg_ar_coef <- DESeq(dds_ar_coef, betaPrior = FALSE)
coef_names <- resultsNames(dds_deg_ar_coef)
coef_interactions <- grep(pattern = "body_part[A-Z]\\.alleleB", x = coef_names, value = TRUE)

#Run the Wald tests for each coefficient (this will compare each coefficient with each other)
#Get all possible contrasts (all combination of elements in 'coef_interactions', 2 at a time)
all_contrasts <- combn(x = c(1:length(coef_interactions)), m = 2)
#Transpose the matrix of contrasts so that it can be read in the for loop
all_contrasts <- t(all_contrasts)

#Run loop for all contrasts of interactions
#Generate empty vector to store all p values
all_pvals_interactions <- c()

for(this_contrast in 1:nrow(all_contrasts)){
  #Select the coefficients for the contrast
  coef_one <- all_contrasts[this_contrast, 1]
  coef_one <- coef_interactions[coef_one]
  coef_two <- all_contrasts[this_contrast, 2]
  coef_two <- coef_interactions[coef_two]
  #Run the contrast through DESeq2
  this_result <- results(dds_deg_ar_coef, contrast = list(coef_one, coef_two))
  #Get p values for that contrast and add to the existing vector
  all_pvals_interactions <- append(all_pvals_interactions, this_result$padj)
}

#Are any of the p values significant (alpga = 0.05)?
which(c < 0.05)
#None
```

Pairwise Wald tests show again that none of the interactions with body part are significant. 

## Plot all significant genes to check variant effect visually

DESeq2 gives a lsit of differentially expressed genes between variants. Some of those are based on very low read counts on average. The following plot can be use to assess visually those differences.

```{r variant_differences_plot, message = FALSE}

#Select significant genes
sig_genes <- rownames(res_deg_ar[which(res_deg_ar$padj < 0.05), ])
counts_specific_genes <- counts(dds_ar, norm = TRUE)[sig_genes, ]


specific_genes <- rep(rownames(counts_specific_genes), ncol(counts_specific_genes))
samples <- rep(colnames(counts_specific_genes), each = nrow(counts_specific_genes))
variant <- gsub(pattern = ".+_", replacement = "", x = samples)
body_part_sp <- gsub(x = samples, pattern = "([0-9]+[BC]?Q?)([A-Z])(_[Bb])", replacement = "\\2")

specific_genes_parsed <- data.frame(c(counts_specific_genes), specific_genes, samples, variant, body_part_sp)

colnames(specific_genes_parsed) <- c("counts", "genes", "samples", "variant", "body_part")

ggplot(data = specific_genes_parsed, aes(x = variant, y = log(counts + 1), fill = body_part)) + geom_violin() + facet_grid(. ~ genes) +
  geom_jitter(position = position_dodge(1))

```

